pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/OneStageAuthenticationState.java

```diff
@@ -20,39 +20,60 @@
- * if completed, what is the AuthRole is.
+ * A class to track single stage authentication. This class assumes that:
+ * 1. {@link #authenticateAsync(AuthData)} is called once and when the {@link CompletableFuture} completes,
+ *    authentication is complete.
+ * 2. Authentication does not expire, so {@link #isExpired()} always returns false.
+ * <p>
+ * See {@link AuthenticationState} for Pulsar's contract on how this interface is used by Pulsar.
  */
 public class OneStageAuthenticationState implements AuthenticationState {
 
-    private final AuthenticationDataSource authenticationDataSource;
-    private final String authRole;
+    private AuthenticationDataSource authenticationDataSource;
+    private final SocketAddress remoteAddress;
+    private final SSLSession sslSession;
+    private final AuthenticationProvider provider;
+    private volatile String authRole;
+
 
+    /**
+     * Constructor for a {@link OneStageAuthenticationState} where there is no authentication performed during
+     * initialization.
+     * @param remoteAddress - remoteAddress associated with the {@link AuthenticationState}
+     * @param sslSession - sslSession associated with the {@link AuthenticationState}
+     * @param provider - {@link AuthenticationProvider} to use to verify {@link AuthData}
+     */
     public OneStageAuthenticationState(AuthData authData,
                                        SocketAddress remoteAddress,
                                        SSLSession sslSession,
-                                       AuthenticationProvider provider) throws AuthenticationException {
-        this.authenticationDataSource = new AuthenticationDataCommand(
-            new String(authData.getBytes(), UTF_8), remoteAddress, sslSession);
-        this.authRole = provider.authenticate(authenticationDataSource);
+                                       AuthenticationProvider provider) {
+        this.provider = provider;
+        this.remoteAddress = remoteAddress;
+        this.sslSession = sslSession;
     }
 
-    public OneStageAuthenticationState(HttpServletRequest request, AuthenticationProvider provider)
-            throws AuthenticationException {
+    public OneStageAuthenticationState(HttpServletRequest request, AuthenticationProvider provider) {
+        // Must initialize this here for backwards compatibility with http authentication
         this.authenticationDataSource = new AuthenticationDataHttps(request);
-        this.authRole = provider.authenticate(authenticationDataSource);
+        this.provider = provider;
+        // These are not used when invoking this constructor.
+        this.remoteAddress = null;
+        this.sslSession = null;
     }
 
     @Override
-    public String getAuthRole() {
+    public String getAuthRole() throws AuthenticationException {
+        if (authRole == null) {
+            throw new AuthenticationException("Must authenticate before calling getAuthRole");
+        }
```

Comment:
This PR looks like a breaking change. In KoP,  this method is used to get the role from the token. After upgrading the dependency, many authorization related tests failed. See https://github.com/streamnative/kop/actions/runs/4343910383/jobs/7586547352

Not sure if it should be reverted in master branch, but it should not be cherry-picked to release branches. @michaeljmarshall 

/cc @Demogorgon314  @nodece @coderzc

Reply:
@BewareMyPower - I agree that we shouldn't cherry pick this line. I don't think I did though, taking a quick look at one of the linked commits. I primarily cherry picked the code that calls `authenticate` on the original auth data. Are you seeing this error in the kop logs?

Reply:
I understand now. That KoP test references pulsar 3.0.0, so it must be from a recent build of master. First, I'll note that I didn't cherry pick the behavior change to old release branches. I cherry-picked the extra authentication call for original auth data, which was code introduced in this PR. Second, I think this change is necessary for enabling PIP 97. I first documented the contract here: https://github.com/apache/pulsar/pull/19283. The primary issue is that creating the `AuthState` should not trigger authenticating the `AuthData`. KoP relied on this implementation detail, and that is why the tests failed. Given that the only impacted parties are plugin developers, I think a PIP, documented release notes, a major version bump, and a fail fast behavior (the object throws an exception), make this change acceptable for 3.0.0. Let me know what you think, thanks.

Reply:
Sorry I missed the fact that you have removed these `cherry-picked` labels.

Regarding this change, I think you should add another constructor to `OneStageAuthenticationState`. Currently, the first argument `authData` is actually not used. Maybe you didn't remove it because it's public. (However, you removed the exception signature) But the semantic changed and it looked weird now.

To keep the compatibility, it's better to mark the constructor as deprecated first and keep it the semantic as "authenticate synchronously". To use the asynchronous semantic, you should add another constructor to do that.

```java
    public OneStageAuthenticationState(SocketAddress remoteAddress,
                                       SSLSession sslSession,
                                       AuthenticationProvider provider) {
        this.provider = provider;
        this.remoteAddress = remoteAddress;
        this.sslSession = sslSession;
    }

    @Deprecated
    public OneStageAuthenticationState(AuthData authData,
                                       SocketAddress remoteAddress,
                                       SSLSession sslSession,
                                       AuthenticationProvider provider) throws AuthenticationException {
        this(remoteAddress, sslSession, provider);
        this.authenticationDataSource = new AuthenticationDataCommand(
                new String(authData.getBytes(), StandardCharsets.UTF_8), remoteAddress, sslSession);
        this.authRole = provider.authenticate(this.authenticationDataSource);
    }
```

> KoP relied on this implementation detail

I'd rather use "semantics" instead of the "implementation detail". Take `Producer#send` as example:
- Semantics: the send is synchronous
- Implementation details: the exception message representation

If you made a change that makes `send` asynchronous, it should be treated as a breaking change. But if you just modified the exception message, for example, adding a common prefix to the exception message, it would be okay.

Reply:
@BewareMyPower - thanks for the context and the suggestion. 

>  I think you should add another constructor to `OneStageAuthenticationState`. 

Thanks for the suggestion, I agree this is the best way to move forward without breaking any other applications when 3.0.0 is released. I also agree that it makes the API clearer because the unused `AuthData` won't be passed in. I shoul be able to get a PR submitted within 24 hours or so.

> (However, you removed the exception signature)

I didn't realize this would cause an issue. I'll add it back when I fix the constructor. Out of curiosity, what is the consequence of breaking source compatibility by removing the exception from the method signature?

Reply:
> Out of curiosity, what is the consequence of breaking source compatibility by removing the exception from the method signature?

No. I just noticed this point when I reviewed again. After looking for some documents and testing the ABI compatibility locally, it seems that removing the exception signature does not affect the ABI compatibility.

Reply:
@BewareMyPower - I finally got a chance to come back to this, and now I remember that I did try to use a similar solution that you proposed while writing this PR, but I found it didn't actually work as you propose. There are a few details to show the issue:

First, the `OneStageAuthenticationState` is not used directly by the broker or by plugins. Instead, the broker and plugins use `AuthenticationProvider#newAuthState` to create the `AuthenticationState` object. Here is the code:

https://github.com/apache/pulsar/blob/fb7f14ceb04d612e456b2e5a834385ae3a97f68f/pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProvider.java#L96-L104

As such, changing the constructor in the way you propose is unlikely to help users that built their own extensions. For example, KoP does the following:

https://github.com/streamnative/kop/blob/c9f66e4d24608dea6c7f4badde1d288b7dc1f74d/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/SchemaRegistryManager.java#L105-L106

Users must rely on the `AuthenticationProvider#newAuthState` in order to make the `AuthenticationProvider` pluggable.

When I noticed this detail while writing this PR, I thought about adding a new method to the `AuthenticationProvider` that would not take the `AuthData` object, in the same way that you proposed for the new `OneStageAuthenticationState` constructor. However, the main issue is how to default that method in the interface. We could use the following definition:

```java
    default AuthenticationState newAuthState(SocketAddress remoteAddress, SSLSession sslSession)
            throws AuthenticationException {
        return new OneStageAuthenticationState(remoteAddress, sslSession, this);
    }
```

This would be analogous to the current `newAuthState` method, but the issue is that custom `AuthenticationProvider` implementations that override the current `newAuthState` method likely don't want to use the `OneStageAuthenticationState`, and that would lead to unexpected behavior.

Another option could be:

```java
    default AuthenticationState newAuthState(SocketAddress remoteAddress, SSLSession sslSession)
            throws AuthenticationException {
        return newAuthState(null, remoteAddress, sslSession);
    }
```

However, that would require breaking the current implementation in the way you're concerned about. (For what it's worth, this option might be the best long term solution for cleaning up the interface if we want to remove the `authData` that is passed. The main point is that it doesn't take care of your concern of not breaking user code.)

Another option:

```java
AuthenticationState newAuthState(SocketAddress remoteAddress, SSLSession sslSession);
```

This unimplemented option has two problems. First, it will break user code if they do not implement the method. Since this whole exercise is an effort to try to prevent users from needing to change any code, that isn't a good option. Second, it removes the `OneStageAuthenticationState` default already present in the interface, which seems problematic.

In my view, the primary challenge is that the `AuthenticationProvider` interface has a brittle default behavior.

Given the above, do you have any ideas on how to make this better for users building their own broker extensions?

Reply:
>  but the issue is that custom AuthenticationProvider implementations that override the current newAuthState method likely don't want to use the OneStageAuthenticationState, and that would lead to unexpected behavior.

Could you show an example of a possible unexpected behavior? IMO, the existing authentication provider only overrides the previous `newAuthState` method still has the previous behavior and is not affected by the new `newAuthState` method.

Reply:
> Could you show an example of a possible unexpected behavior? IMO, the existing authentication provider only overrides the previous `newAuthState` method still has the previous behavior and is not affected by the new `newAuthState` method.

The challenge comes from the fact that Apache Pulsar will transition to using the new method to build the state object and if that method is not overridden by the custom `AuthenticationProvider` implementation, the `OneStageAuthenticationState` will be used. That will lead to unexpected behavior.

A good example is the `AuthenticationProviderToken` class. That one would "work" with the `OneStageAuthenticationState` with the key difference being the `isExpired` value from `OneStageAuthenticationState` always returns `false` while the `TokenAuthenticationState` returns `true` when the token's `exp` claim indicates the token has expired.

Reply:
I got it. The key point is that Pulsar will prefer the new overload but the existing implementations might still implement the old overload. It seems that there is no better way to solve it perfectly.